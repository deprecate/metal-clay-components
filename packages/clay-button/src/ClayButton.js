import 'clay-icon';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayButton.soy.js';

/**
 * Metal Clay Button component.
 */
class ClayButton extends ClayComponent {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayButton.STATE = {
	/**
	 * Aria label attribute for the button element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	ariaLabel: Config.string(),

	/**
	 * Renders the button as a block element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?boolean}
	 * @default false
	 */
	block: Config.bool().value(false),

	/**
	 * Data to add to the element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?object}
	 * @default undefined
	 */
	data: Config.object(),

	/**
	 * The button disabled attribute.
	 * @instance
	 * @memberof ClayButton
	 * @type {?boolean}
	 * @default false
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Render ClayIcon in the ClayButton element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?Object|undefined}
	 * @default undefined
	 */
	icon: Config.shapeOf({
		alignment: Config.oneOf(['left', 'right']),
		spritemap: Config.string().required(),
		symbol: Config.string().required(),
	}),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * The label of the button content.
	 * @instance
	 * @memberof ClayButton
	 * @type {?html|string|undefined}
	 * @default undefined
	 */
	label: Config.any(),

	/**
	 * Flag to indicate if button should be monospaced.
	 * @instance
	 * @memberof ClayButton
	 * @type {?bool}
	 * @default false
	 */
	monospaced: Config.bool().value(false),

	/**
	 * The name attribute value of the element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	name: Config.string(),

	/**
	 * The name that will make the CSS class name that will represent the
	 * button size.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	size: Config.oneOf(['sm']),

	/**
	 * The css class that colors the button.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default primary
	 */
	style: Config.oneOf([
		'borderless',
		'link',
		'primary',
		'secondary',
		'unstyled',
	]).value('primary'),

	/**
	 * The type attribute value of the element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	type: Config.oneOf(['button', 'reset', 'submit']),

	/**
	 * The value attribute value of the element.
	 * @instance
	 * @memberof ClayButton
	 * @type {?string|undefined}
	 * @default undefined
	 */
	value: Config.string(),
};

defineWebComponent('clay-button', ClayButton);

Soy.register(ClayButton, templates);

export {ClayButton};
export default ClayButton;
