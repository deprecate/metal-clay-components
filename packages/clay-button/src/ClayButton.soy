{namespace ClayButton}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param? ariaLabel: string}
	{@param? block: bool}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? icon: [
		alignment: string,
		monospaced: bool,
		spritemap: string,
		symbol: string
	]}
	{@param? id: string}
	{@param? label: html|string}
	{@param? name: string}
	{@param? size: string}
	{@param? style: string}
	{@param? type: string}
	{@param? value: string}

	{let $attributes kind="attributes"}
		{call .attributes}
			{param ariaLabel: $ariaLabel /}
			{param block: $block /}
			{param disabled: $disabled /}
			{param elementClasses: $elementClasses /}
			{param icon: $icon /}
			{param id: $id /}
			{param label: $label /}
			{param name: $name /}
			{param size: $size /}
			{param style: $style /}
			{param type: $type /}
			{param value: $value /}
		{/call}
	{/let}

	<button {$attributes}>
		{call .content}
			{param icon: $icon /}
			{param label: $label /}
		{/call}
	</button>
{/template}

/**
 * Renders element content
 */
{template .content}
	{@param? icon: [
		alignment: string,
		monospaced: bool,
		spritemap: string,
		symbol: string
	]}
	{@param? label: html|string}

	{let $iconAlignment: $icon and $icon.alignment ? $icon.alignment : 'left' /}

	{if $icon and $iconAlignment == 'left'}
		{call .icon}
			{param icon: $icon /}
		{/call}
	{/if}

	{$label ?: ''}

	{if $icon and $iconAlignment == 'right'}
		{call .icon}
			{param icon: $icon /}
		{/call}
	{/if}
{/template}

/**
 * Renders icon element
 */
{template .icon}
	{@param? icon: [
		alignment: string,
		monospaced: bool,
		spritemap: string,
		symbol: string
	]}

	{call ClayIcon.render}
		{param monospaced: $icon.monospaced /}
		{param spritemap: $icon.spritemap /}
		{param symbol: $icon.symbol /}
	{/call}
{/template}

/**
 * Renders html attributes for button element
 */
{template .attributes kind="attributes"}
	{@param? ariaLabel: string}
	{@param? block: bool}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? icon: [
		alignment: string,
		monospaced: bool,
		spritemap: string,
		symbol: string
	]}
	{@param? id: string}
	{@param? label: string}
	{@param? name: string}
	{@param? size: string}
	{@param? style: string}
	{@param? type: string}
	{@param? value: string}

	{let $classes kind="text"}
		btn
		{if $block}
			{sp}btn-block
		{/if}
		{if $elementClasses}
			{sp}{$elementClasses}
		{/if}
		{if $size}
			{sp}btn-{$size}
		{/if}
		{if $style}
			{sp}btn-{$style}
		{else}
			{sp}btn-primary
		{/if}
	{/let}

	class="{$classes}"

	{if $ariaLabel}
		aria-label="{$ariaLabel}"
	{elseif $label}
		{$label}
	{elseif $icon}
		{$icon.symbol}
	{/if}

	{if $disabled}
		disabled
	{/if}

	{if $id}
		id="{$id}"
	{/if}

	{if $name}
		name="{$name}"
	{/if}

	{if $value}
		value="{$value}"
	{/if}

	{if $type}
		type="{$type}"
	{/if}
{/template}
