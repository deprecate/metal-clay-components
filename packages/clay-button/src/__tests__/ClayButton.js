import ClayButton from '../ClayButton';

let button;

describe('ClayButton', function() {
  afterEach(() => {
    if (button) {
      button.dispose();
    }
  });

  it('should generate the default markup', function() {
    button = new ClayButton();

    expect(button).toMatchSnapshot();
  });

  it('should render a submit button', function() {
    button = new ClayButton({
      type: 'submit',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a disabled button', function() {
    button = new ClayButton({
      disabled: true,
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a secondary button', function() {
    button = new ClayButton({
      style: 'secondary',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with label', function() {
    button = new ClayButton({
      label: 'Label',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with name', function() {
    button = new ClayButton({
      name: 'myButton',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with label and ariaLabel', function() {
    button = new ClayButton({
      ariaLabel: 'My Description',
      label: 'Visit Liferay.com',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a small button', function() {
    button = new ClayButton({
      size: 'sm',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a block button', function() {
    button = new ClayButton({
      block: true,
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with icon', function() {
    button = new ClayButton({
      icon: {
        spritemap: 'icons.svg',
        symbol: 'plus',
      },
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with icon and monospaced true', function() {
    button = new ClayButton({
      icon: {
        spritemap: 'icons.svg',
        symbol: 'plus',
      },
      monospaced: true,
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with icon and label', function() {
    button = new ClayButton({
      icon: {
        spritemap: 'icons.svg',
        symbol: 'plus',
      },
      label: 'Label',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with label and icon', function() {
    button = new ClayButton({
      icon: {
        alignment: 'right',
        spritemap: 'icons.svg',
        symbol: 'plus',
      },
      label: 'Label',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with id', function() {
    button = new ClayButton({
      id: 'myButton',
    });

    expect(button).toMatchSnapshot();
  });

  it('should render a button with custom classes', function() {
    button = new ClayButton({
      elementClasses: 'my-custom-class',
    });

    expect(button).toMatchSnapshot();
  });
});
