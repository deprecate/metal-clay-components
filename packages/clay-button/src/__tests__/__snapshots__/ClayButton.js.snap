// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayButton should generate the default markup 1`] = `<button class="btn btn-primary"></button>`;

exports[`ClayButton should render a block button 1`] = `<button class="btn btn-block btn-primary"></button>`;

exports[`ClayButton should render a button with icon 1`] = `
<button class="btn btn-primary">
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus icon-monospaced">
    <use xlink:href="icons.svg#plus"></use>
  </svg>
</button>
`;

exports[`ClayButton should render a button with icon and label 1`] = `
<button class="btn btn-primary">
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus icon-monospaced">
    <use xlink:href="icons.svg#plus"></use>
  </svg>Label</button>
`;

exports[`ClayButton should render a button with id 1`] = `<button class="btn btn-primary" id="myButton"></button>`;

exports[`ClayButton should render a button with label 1`] = `<button class="btn btn-primary">Label</button>`;

exports[`ClayButton should render a button with label and ariaLabel 1`] = `<button class="btn btn-primary" aria-label="My Description">Visit Liferay.com</button>`;

exports[`ClayButton should render a button with label and icon 1`] = `
<button class="btn btn-primary">Label
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus icon-monospaced">
    <use xlink:href="icons.svg#plus"></use>
  </svg>
</button>
`;

exports[`ClayButton should render a button with name 1`] = `<button class="btn btn-primary" name="myButton"></button>`;

exports[`ClayButton should render a disabled button 1`] = `<button class="btn btn-primary" disabled=""></button>`;

exports[`ClayButton should render a secondary button 1`] = `<button class="btn btn-secondary"></button>`;

exports[`ClayButton should render a small button 1`] = `<button class="btn btn-sm btn-primary"></button>`;

exports[`ClayButton should render a submit button 1`] = `<button class="btn btn-primary" type="submit"></button>`;
